### PR TITLE
feat: adding multiparty sharing pattern

### PR DESCRIPTION
**Pull request by 0xng**
# 🤖 Linear
Closes AZT-107

### CODE REVIEW

File: patterns/multiparty-note-sharing/broadcaster/src/main.nr [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+
+        // If the sender is not the owner, ensure the caller is authorized to call this contract.
+        // For users that have notes where the owner is a contract, this forces to go through the contract to nullify the note.
+        if(!sender.eq(shared_note.owner)) {
+            assert_current_call_valid_authwit(&mut context, shared_note.owner);
+        } else {
+            assert(nonce == 0, "Invalid nonce");
+        }
+
+        // This extra constraint is needed to avoid a compiler bug
```
Comment (simon-something) [inline patterns/multiparty-note-sharing/broadcaster/src/main.nr]: "bug" is maybe a bit too opiniated? Is it working if we include the previous check in the if block? Maybe tidier to then have 
```
if owner != 0:
  authwit?
  etc
else "does not exist"
```

File: patterns/multiparty-note-sharing/broadcaster/src/test/e2e.test.ts [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    return pxe;
+  };
+  
+  // Setup: Set the sandbox
+  beforeAll(async () => {
+    pxe = await setupSandbox();
+    [user, user1, deployer] = await getInitialTestAccountsWallets(pxe);
+    user2 = await createAccount(pxe);
+    user3 = await createAccount(pxe);
+    user4 = await createAccount(pxe);
```
Comment (simon-something) [inline patterns/multiparty-note-sharing/broadcaster/src/test/e2e.test.ts:45]: maybe worth noting somewhere that we have the pk in the pxe due to this createAccount(...)

File: patterns/multiparty-note-sharing/broadcaster/src/types/shared_note.nr [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    }
+
+    /// @notice Sets the note header.
+    /// @param  header: the header to set
+    pub fn set_header(&mut self, header: NoteHeader) {
+        self.header = header;
+    }
+
+    /// @notice broadcasts the note to the owner as long as the owner is not a contract.
+    /// @dev    it's not emitted to contract owners as they don't have encryption keys.
```
Comment (simon-something) [inline patterns/multiparty-note-sharing/broadcaster/src/types/shared_note.nr]: "contract owner" has a bit of a double meaning, "contract owning a note" or similar

File: patterns/multiparty-note-sharing/broadcaster/src/main.nr (hunk 1 above)
Comment (0xng) [reply patterns/multiparty-note-sharing/broadcaster/src/main.nr]: Changed `bug` for `error` and added a bit of context, but left the checks separated

**Review by simon-something — APPROVED** [review]

**Pull request merged**